### PR TITLE
Add extend option to ActiveRecord named scopes

### PR DESCRIPTION
**Pull request by Bodacious**
### Motivation / Background

Make it possible to extend an ActiveRecord scope by one or more module as well as using the existing block definition. 

e.g.

``` ruby 
class User < ApplicationRecord 
  scope :active, -> { where(status: "active") }, extend: ActiveUserMethods
end
``` 

### Detail

Currently, one can extend the `ActiveRecord::Relation` returned from an `ActiveRecord` scope method by passing a block with extra method definitions:

``` ruby 
class Product < ActiveRecord::Base
  scope :free, -> { where(price_cents: 0) } do
    def sum_price_cents = 0
    def ave_price_cents = 0
  end

  def self.sum_price_cents
    sum(:price_cents)
  end

  def self.ave_price_cents
    average(:price_cents)
  end
end

# Queries the DB
Product.sum_price_cents  # => 12345
# Doesn't Query the DB; returns 0
Product.free.sum_price_cents # => 0
```

The block method shown above is limited, though, as:
1. It doesn't allow one to share behaviour between scopes. 
2. Adding more methods means blocks can start to bloat and become unsightly. 
3. Finally, `has_many` also supports this block syntax, but supports an additional `:extend` option to pass a Module or Array of Modules. Adding support for this option in the `scope` method helps to make the `ActiveRecord` API more consistent, and reduce surprises when working with `ActiveRecord`.

This PR adds support for an `:extend` option:

``` ruby 
module FreeProductCalculations 
  def sum_price_cents = 0
  def ave_price_cents = 0
end

class Product < ActiveRecord::Base
  scope :free, -> { where(price_cents: 0) }, extend: FreeProductCalculations

  def self.sum_price_cents
    sum(:price_cents)
  end

  def self.ave_price_cents
    average(:price_cents)
  end
end
```

### Additional information

I've benchmarked these two solutions, and the difference is negligible:

```
Warming up --------------------------------------
              Extend     3.369k i/100ms
               Block     3.001k i/100ms
Calculating -------------------------------------
              Extend     32.570k (± 5.7%) i/s -    165.081k in   5.086707s
               Block     33.212k (± 3.2%) i/s -    168.056k in   5.065352s
```

[Link to full benchmark.](https://gist.github.com/Bodacious/d6e652cccb630a3613a0838f2a2dc004)

### Checklist

Before submitting the PR make sure the following are checked:

* [x] This Pull Request is related to one change. Changes that are unrelated should be opened in separate PRs.
* [x] Commit message has a detailed description of what changed and why. If this PR fixes a related issue include it in the commit message. Ex: `[Fix #issue-number]`
* [x] Tests are added or updated if you fix a bug or add a feature.
* [x] CHANGELOG files are updated for the changed libraries if there is a behavior change or additional feature. Minor bug fixes and documentation changes should not be included.

### CODE REVIEW

**Comment by MatheusRich** [comment]
Would it be possible to use a proc/lambda to get a reusable block? Something like

```rb
some_reusable_block = ->(obj) { "#{obj}!" }
%w[a b c].map(&some_reusable_block)
=> ["a!", "b!", "c!"]
```

You could extract it to a constant and to a different file if you want to.
Your example could look like this:

```rb
# Possibly in a different file
FREE_PRODUCT_CALCULATIONS = -> do
  def sum_price_cents = 0
  def ave_price_cents = 0
end

class Product < ActiveRecord::Base
  scope :free, -> { where(price_cents: 0) }, &FREE_PRODUCT_CALCULATIONS
end
```

I'm not sure how faster/slower that would be.

File: activerecord/test/cases/scoping/named_scoping_test.rb [hunk 1]
```diff
@@ -626,4 +626,80 @@ def test_scope_with_annotation
...
+    end
+    klass.scope :with_both_extend_and_block, -> { }, extend: extending_module do
+      def block_works?
+        true
+      end
+    end
+
+    assert_predicate klass.with_both_extend_and_block, :extend_works?
+    assert_predicate klass.with_both_extend_and_block, :block_works?
+  end
```
Comment (MatheusRich) [inline activerecord/test/cases/scoping/named_scoping_test.rb:691]: Can we add a second case here to demonstrate which one (the block or the module) has precedence?

Or maybe we could just modify this existing test
```suggestion  def test_scope_with_extend_option_and_block
    klass = Class.new(ActiveRecord::Base)
    klass.table_name = "topics"
    extending_module = Module.new do
      def block_works?
        false
      end
      
      def extend_works?
        true
      end
    end
    klass.scope :with_both_extend_and_block, -> { }, extend: extending_module do
      def block_works?
        true
      end
    end

    assert_predicate klass.with_both_extend_and_block, :extend_works?
    assert_predicate klass.with_both_extend_and_block, :block_works?
  end```

**Comment by skipkayhil** [comment]
Hey @Bodacious, thanks for the PR!

Are you not able to use `include` in the scope block already?

```ruby
# frozen_string_literal: true

require "bundler/inline"

gemfile(true) do
  source "https://rubygems.org"

  git_source(:github) { |repo| "https://github.com/#{repo}.git" }

  # Activate the gem you are reporting the issue against.
  gem "activerecord", "~> 7.0.0"
  gem "sqlite3"
end

require "active_record"
require "minitest/autorun"
require "logger"

# This connection will do for database-independent bug reports.
ActiveRecord::Base.establish_connection(adapter: "sqlite3", database: ":memory:")
ActiveRecord::Base.logger = Logger.new(STDOUT)

ActiveRecord::Schema.define do
  create_table :posts, force: true do |t|
    t.boolean :active
  end
end

module MoreMethods
  def more_testing
    puts 'more testing'
  end
end

class Post < ActiveRecord::Base
  scope :active, -> { where(active: true) } do
    include MoreMethods

    def testing
      puts 'testing'
    end
  end
end

class BugTest < Minitest::Test
  def test_scope_includes
    assert !Post.all.respond_to?(:testing)
    assert Post.active.respond_to?(:testing)
  end

  def test_scope_more_includes
    assert !Post.all.respond_to?(:more_testing)
    assert Post.active.respond_to?(:more_testing)
  end
end
```

**Comment by Bodacious** [comment]
> Would it be possible to use a proc/lambda to get a reusable block? Something like
> 
> ```ruby
> some_reusable_block = ->(obj) { "#{obj}!" }
> %w[a b c].map(&some_reusable_block)
> => ["a!", "b!", "c!"]
> ```
> 
> You could extract it to a constant and to a different file if you want to. Your example could look like this:
> 
> ```ruby
> # Possibly in a different file
> FREE_PRODUCT_CALCULATIONS = -> do
>   def sum_price_cents = 0
>   def ave_price_cents = 0
> end
> 
> class Product < ActiveRecord::Base
>   scope :free, -> { where(price_cents: 0) }, &FREE_PRODUCT_CALCULATIONS
> end
> ```
> 
> I'm not sure how faster/slower that would be.

Thanks @MatheusRich. I actually did something very similar in my app, but felt that this pattern was a bit uglier than it needed to be (see below). It also adds an additional anonymous module to the Relation object, but the proposed approach in my PR does not ([that's demonstrated in this Gist](https://gist.github.com/Bodacious/7956303c847e9666b2426f72ab2d2425)). It's not the end of the world, but it's probably better to avoid creating unnecessary modules when we can.

My solution is below, for anyone who is curious 🤓 

---

``` ruby 
class Product < ActiveRecord::Base
  
  scope :free, -> { where(price_cents: 0) }, &FreeProductCalculations

end

module FreeProductCalculations
  def self.to_proc 
    Proc.new { include FreeProductCalculations }
  end
end
```

**Comment by Bodacious** [comment]
> Hey @Bodacious, thanks for the PR!
> 
> Are you not able to use `include` in the scope block already?

Thanks @skipkayhil. Yes, include within the block also works, but has the same problem that I've described above—it includes an extra module within the Relation that is redundant.

Is _is_ possible to use [`extending`](https://api.rubyonrails.org/classes/ActiveRecord/QueryMethods.html#method-i-extending) within a scope's definition, and this does _not_ add an extra module (my PR uses this behind the scenes). But I feel that this method is not as well-known, and felt that matching the behaviour of `has_many` would be a small improvement.

**Comment by rafaelfranca** [comment]
Thank you for the pull request but using `extending` is the way to go.

**Comment by zzak** [comment]
> But I feel that this method is not as well-known, and felt that matching the behaviour of has_many would be a small improvement.

If this is the case, then perhaps that is an opportunity to improve the documentation :bow:

**Pull request closed without merging**